codec-http/src/main/java/io/netty/handler/codec/http/HttpKeepAliveHandler.java

```diff
@@ -0,0 +1,135 @@
+ * before any other handlers that might send a {@link HttpResponse}.
+ * <blockquote>
+ * <pre>
+ *  {@link ChannelPipeline} p = ...;
+ *  ...
+ *  p.addLast("serverCodec", new {@link HttpServerCodec}());
+ *  p.addLast("httpKeepAlive", <b>new {@link HttpKeepAliveHandler}()</b>);
+ *  p.addLast("aggregator", new {@link HttpObjectAggregator}(1048576));
+ *  ...
+ *  p.addLast("handler", new HttpRequestHandler());
+ *  </pre>
+ * </blockquote>
+ */
+@ChannelHandler.Sharable
+public class HttpKeepAliveHandler extends ChannelDuplexHandler {
+
+    private static final String MULTIPART_PREFIX = "multipart";
+
+    /**
+     * Will close the channel if the connection couldn't be persisted.
+     * <p>
+     * Only works if HttpKeepAliveHandler is part of the pipeline.
+     */
+    private static final ChannelFutureListener CLOSE_UNLESS_KEEP_ALIVE = new ChannelFutureListener() {
+        @Override
+        public void operationComplete(ChannelFuture channelFuture) throws Exception {
+            Boolean keepAlive = channelFuture.channel().attr(KEEP_ALIVE_ATTR).get();
+            if (!TRUE.equals(keepAlive) && channelFuture.channel().isActive()) {
+                channelFuture.channel().close();
+            }
+        }
+    };
+
+    private static final AttributeKey<Boolean> KEEP_ALIVE_ATTR =
+            AttributeKey.valueOf(HttpKeepAliveHandler.class, "KEEP_ALIVE_ATTR");
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        // read message and track if it was keepAlive
+        if (msg instanceof HttpRequest) {
+            HttpRequest request = (HttpRequest) msg;
+            ctx.channel().attr(KEEP_ALIVE_ATTR).set(isKeepAlive(request));
+        }
+        super.channelRead(ctx, msg);
+    }
+
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
+        // modify message on way out to add headers if needed
+        if (msg instanceof HttpResponse) {
+            HttpResponse response = (HttpResponse) msg;
+            // Only mess with connection header if not already set
+            Attribute<Boolean> keepAliveRequested = ctx.channel().attr(KEEP_ALIVE_ATTR);
+            if (response.headers().contains(HttpHeaderNames.CONNECTION)) {
+                keepAliveRequested.set(isKeepAlive(response));
+            } else {
+                if (isKeepAliveEligible(response, keepAliveRequested)) {
+                    setKeepAlive(response, true);
+                } else {
+                    setKeepAlive(response, false);
+                    // No longer keep alive as the client can't tell when the message is done unless we close connection
+                    keepAliveRequested.set(FALSE);
+                }
+            }
+        }
+        if (msg instanceof LastHttpContent) {
+            promise.addListener(CLOSE_UNLESS_KEEP_ALIVE);
```

Comment:
Is it possible the connection will be prematurely closed when pipelineing is being used? For example if this handler is used on the server, the client issues 2 pipelined requests with the second request including a `Connection: Close` header, and the server reads both requests before sending any responses. In this case won't we close the connection after the server writes the first response?